controllers/util/solr_tls_util.go

```diff
@@ -27,171 +27,217 @@ import (
-	mainContainer.Env = append(mainContainer.Env, tls.envVars()...)
 
-	if tls.Options.PKCS12Secret != nil {
-		// the TLS files are mounted from a secret, setup the volumes and mounts
-		template.Spec.Volumes = append(template.Spec.Volumes, tls.volumes()...)
-		mainContainer.VolumeMounts = append(mainContainer.VolumeMounts, tls.volumeMounts()...)
-
-		// track the MD5 of the TLS cert (from secret) to trigger restarts if the cert changes
-		if tls.Options.RestartOnTLSSecretUpdate && tls.CertMd5 != "" {
-			if template.Annotations == nil {
-				template.Annotations = make(map[string]string, 1)
-			}
-			template.Annotations[SolrTlsCertMd5Annotation] = tls.CertMd5
-		}
+	// the TLS files are mounted from a secret, setup the volumes and mounts
+	vols, mounts := tls.volumesAndMounts(namePrefix)
+
+	// We need an initContainer to convert a TLS cert into the pkcs12 format Java wants (using openssl)
+	// but openssl cannot write to the /var/solr/tls directory because of the way secret mounts work
+	// so we need to mount an empty directory to write pkcs12 keystore into
+	if tls.NeedsPkcs12InitContainer {
+		vols = append(vols, corev1.Volume{Name: "pkcs12", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}})
+		mounts = append(mounts, corev1.VolumeMount{Name: "pkcs12", ReadOnly: false, MountPath: DefaultWritableKeyStorePath})
+		pkcs12InitContainer := tls.generatePkcs12InitContainer(mainContainer.Image, mainContainer.ImagePullPolicy, mounts)
+		template.Spec.InitContainers = append(template.Spec.InitContainers, pkcs12InitContainer)
+	}
+	template.Spec.Volumes = append(template.Spec.Volumes, vols...)
+	mainContainer.VolumeMounts = append(mainContainer.VolumeMounts, mounts...)
 
-		// use an initContainer to convert the keystore to p12 format if needed
-		if tls.NeedsPkcs12InitContainer {
-			pkcs12InitContainer := tls.generatePkcs12InitContainer(mainContainer.Image, mainContainer.ImagePullPolicy)
-			template.Spec.InitContainers = append(template.Spec.InitContainers, pkcs12InitContainer)
+	// track the MD5 of the TLS cert (from secret) to trigger restarts if the cert changes
+	if tls.Options.RestartOnTLSSecretUpdate && tls.CertMd5 != "" {
+		if template.Annotations == nil {
+			template.Annotations = make(map[string]string, 1)
 		}
+		template.Annotations[tls.CertMd5Annotation] = tls.CertMd5
 	}
 
 	return mainContainer
 }
 
 // Get a list of volumes for the keystore and optionally a truststore loaded from a TLS secret
-func (tls *TLSConfig) volumes() []corev1.Volume {
+func (tls *TLSConfig) volumesAndMounts(namePrefix string) ([]corev1.Volume, []corev1.VolumeMount) {
 	optional := false
 	defaultMode := int32(0664)
-	vols := []corev1.Volume{
-		{
-			Name: "keystore",
+	vols := []corev1.Volume{}
+	mounts := []corev1.VolumeMount{}
+	keystoreSecretName := ""
+
+	opts := tls.Options
+	if opts.PKCS12Secret != nil {
+		keystoreSecretName = opts.PKCS12Secret.Name
+		volName := namePrefix + "keystore"
```

Comment:
maybe "-keystore"?

Reply:
the prefix is either "" or "client-" ... I'll fix it with a helper method on `tls.Options`

Reply:
Oh then this is fine, ignore